atd-vze/src/views/Users/Users.js

```diff
@@ -66,28 +68,82 @@ const Users = () => {
 
   const [userList, setUserList] = useState(null);
   const [totalUsers, setTotalUsers] = useState(null);
-  const [page, setPage] = useState(0);
+  const [currentPage, setCurrentPage] = useState(0);
+  const [allUsers, setAllUsers] = useState([]);
+  const [copyUserEmailsClicked, setCopyUserEmailsClicked] = useState(false);
   const perPage = 50;
 
+  // Sets the user list and the total number of users for the current page
   useEffect(() => {
+    getCurrentPageUsers(currentPage, perPage).then(res => {
+      setUserList(res.data.users);
+      setTotalUsers(res.data.total);
+    });
+    // eslint-disable-next-line
+  }, [token, currentPage, perPage]);
+
+  // Handles the copy user emails button click
+  useEffect(() => {
```

Comment:
I was looking at this `useEffect` with fresh eyes, and I'm seeing what @johnclary was suggesting earlier about tying these events to the click handler of the copy button. What do you think about something like below? I think that `getAllUsers` could also be moved outside of the component definition too.

The `isFetchingAllUsers` state is total pseudocode so I'm not sure if that is an addition or maybe `copyUserEmailsClicked` could fit in here to tracking loading.

```js
 const handleCopyUserEmails = async () => {
   if(allUsers === null) {
       setIsFetchingAllUsers(true);
       const users = await getAllUsers(perPage, pageCount);
       setAllUsers(users);
    }
    getUserEmails(users);
    setIsFetchingAllUsers(false);
  };
```
The hard part of this is that the button click event depends on the initial data fetch for the total number of user records to iterate in `getAllUsers`. I think that could be addressed by disabling this button until the first page of users is returned.

Reply:
nice—i really like this example! letting the button click trigger the fetching allows the state logic to be a lot tighter. this looks like the way to go to me.